### PR TITLE
Return invalid username error instead of unknown error

### PR DESCRIPTION
**Pull request by SaurabhMSFT**
If the username is empty or malformed then a InvalidUsername error is returned instead of UnknownError

Corresponding PR in common repo: https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/2297

### CODE REVIEW

File: msal/src/main/java/com/microsoft/identity/nativeauth/statemachine/errors/SignInErrors.kt [hunk 1]
```diff
@@ -39,6 +39,8 @@ open class SignInError(
     fun isUserNotFound(): Boolean = this.errorType == ErrorTypes.USER_NOT_FOUND
 
     fun isInvalidCredentials(): Boolean = this.errorType == SignInErrorTypes.INVALID_CREDENTIALS
+
+    fun isInvalidUsername(): Boolean = this.errorType == ErrorTypes.INVALID_USERNAME
```
Comment (Yuki-YuXin) [inline msal/src/main/java/com/microsoft/identity/nativeauth/statemachine/errors/SignInErrors.kt:43]: Do we need a related sample app update PR?

File: msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationJavaTest.java [hunk 2]
```diff
@@ -2654,9 +2650,9 @@ public void testSignUpEmptyUsernameNoException() throws ExecutionException, Inte
         SignUpResult signUpResult = signUpTestCallback.get();
 
         assertTrue(signUpResult instanceof SignUpError);
+        assertTrue(((SignUpError) signUpResult).isInvalidUsername());
```
Comment (rmccahill) [inline msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationJavaTest.java:2649]: It looks like the assertion here is swapped from false to true - is that due to the old testing logic being incorrect?

File: msal/src/main/java/com/microsoft/identity/nativeauth/statemachine/errors/SignInErrors.kt (hunk 1 above)
Comment (SaurabhMSFT) [reply msal/src/main/java/com/microsoft/identity/nativeauth/statemachine/errors/SignInErrors.kt:43]: No update to sample app repo is required.

File: msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationJavaTest.java (hunk 2 above)
Comment (SaurabhMSFT) [reply msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationJavaTest.java:2649]: No, previously the error returned was UnknownError so checking it was not invalidUsername was correct.

File: msal/src/main/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplication.kt [hunk 3]
```diff
@@ -357,6 +365,15 @@ class NativeAuthPublicClientApplication(
                             channel = result.challengeChannel
                         )
                     }
+                    is INativeAuthCommandResult.InvalidUsername -> {
```
Comment (SammyO) [inline msal/src/main/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplication.kt:368]: See comment in https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/2297/files#r1466685798 - in the commits in common right now, this result is never returned.

**Review by SammyO — CHANGES_REQUESTED** [review]

File: msal/src/main/java/com/microsoft/identity/nativeauth/statemachine/errors/SignUpErrors.kt [hunk 4]
```diff
@@ -38,11 +38,6 @@ internal class SignUpErrorTypes {
          */
         const val USER_ALREADY_EXISTS = "user_already_exists"
 
-        /* The INVALID_USERNAME value indicates the username provided by the user is not acceptable to the server.
```
Comment (SammyO) [inline msal/src/main/java/com/microsoft/identity/nativeauth/statemachine/errors/SignUpErrors.kt:41]: So we're combining the SDK generated invalid username error with the API generated one? I'm in favour of that, but this requires a cleanup in common: `SignUpCommandResult.InvalidEmail` result.

**Review by SammyO — CHANGES_REQUESTED** [review]

File: msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationKotlinTest.kt [hunk 5]
```diff
@@ -1276,7 +1277,7 @@ class NativeAuthPublicClientApplicationKotlinTest : PublicClientApplicationAbstr
 
         val result = application.signIn(emptyString, password)
         assertTrue(result is SignInError)
-        assertTrue((result as SignInError).errorType == null)
```
Comment (SammyO) [inline msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationKotlinTest.kt:1279]: I don't understand how this test passed in the past. A ClientException should have been thrown, but that's not in the @Test annotation 🤔

Comment (SammyO) [reply msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationKotlinTest.kt:1279]: note: the `@throws` can be removed from the comments

File: msal/src/main/java/com/microsoft/identity/nativeauth/statemachine/errors/SignUpErrors.kt (hunk 4 above)
Comment (SaurabhMSFT) [reply msal/src/main/java/com/microsoft/identity/nativeauth/statemachine/errors/SignUpErrors.kt:41]: I have cleaned it up.

File: msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationKotlinTest.kt (hunk 5 above)
Comment (SaurabhMSFT) [reply msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationKotlinTest.kt:1279]: Done

File: msal/src/main/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplication.kt (hunk 3 above)
Comment (SaurabhMSFT) [reply msal/src/main/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplication.kt:368]: Updated

**Review by Yuki-YuXin — APPROVED** [review]

**Review by rmccahill — APPROVED** [review]

File: msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationKotlinTest.kt (hunk 5 above)
Comment (SammyO) [reply msal/src/test/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplicationKotlinTest.kt:1279]: Ah, when using Java SDK methods the SDK wraps exceptions that are thrown into `callback.onError(e)`.

**Review by SammyO — APPROVED** [review]

File: msal/src/main/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplication.kt [hunk 6]
```diff
@@ -491,6 +508,14 @@ class NativeAuthPublicClientApplication(
                 )
             }
 
+            if (username.isBlank()) {
+                return@withContext SignUpError(
+                    errorType = ErrorTypes.INVALID_USERNAME,
+                    errorMessage = "Empty or blank username",
+                    correlationId = "UNSET"   //TODO use null correlation id here and in other places in this file instead of UNSET
```
Comment (SammyO) [inline msal/src/main/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplication.kt]: Why this TODO? Why can't be left UNSET?

Comment (SaurabhMSFT) [reply msal/src/main/java/com/microsoft/identity/nativeauth/NativeAuthPublicClientApplication.kt]: This was old TODO when `correlationId` was nullable in your PR. This has now been removed.

**Pull request merged**